core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java

```diff
@@ -0,0 +1,154 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.hydrator.plugin.batch.action;
+
+import ch.ethz.ssh2.Connection;
+import ch.ethz.ssh2.Session;
+import ch.ethz.ssh2.StreamGobbler;
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.plugin.PluginConfig;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+
+import co.cask.cdap.etl.api.action.Action;
+import co.cask.cdap.etl.api.action.ActionContext;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Charsets;
+import com.google.common.base.Strings;
+import com.google.common.io.CharStreams;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import javax.annotation.Nullable;
+
+/**
+ * SSH into a remote machine and execute a script on that machine.
+ * A user must specify username and keypair authentication credentials.
+ * Optionals include port and machine URL
+ */
+@Plugin(type = Action.PLUGIN_TYPE)
+@Name("SSHAction")
+@Description("Action to run a script on remote machine")
+public class SSHAction extends Action {
+  private final SSHActionConfig config;
+
+  private static final Logger LOG = LoggerFactory.getLogger(SSHAction.class);
+
+  public SSHAction(SSHActionConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void run(final ActionContext context) throws Exception {
+    Connection connection = new Connection(config.host);
+    Session session = connection.openSession();
+    try {
+      connection.connect();
+
+      String password = Strings.isNullOrEmpty(config.password) ? null : config.password;
+      if (!connection.authenticateWithPublicKey(config.user, new File(config.privateKeyFile), password)) {
+        throw new IOException(String.format("SSH authentication error when connecting to %s@%s on port %d",
+                                            config.user, config.host, config.port));
+      }
+
+      session.execCommand(config.cmd);
+
+      InputStream stdout = new StreamGobbler(session.getStdout());
+      BufferedReader outBuffer = new BufferedReader(new InputStreamReader(stdout, Charsets.UTF_8));
+      InputStream stderr = new StreamGobbler(session.getStderr());
+      BufferedReader errBuffer = new BufferedReader(new InputStreamReader(stderr, Charsets.UTF_8));
+
+      String out = CharStreams.toString(outBuffer);
+      String err = CharStreams.toString(errBuffer);
+
+      LOG.error("Errors: {}", err);
+
+      if (config.output != null) {
+        context.getArguments().set(config.output, out);
+      }
+    } finally {
+      connection.close();
+      session.close();
```

Comment:
oh actually the docs say "Close the connection to the SSH-2 server. All assigned sessions will be closed, too." Missed that. So then we don't even need to specify `session.close()`